test/e2e/storage/pvc_storageclass.go

```diff
@@ -18,172 +18,168 @@ package storage
-		framework.ExpectNoError(err)
-		for _, sc := range classes.Items {
-			if _, found := changedClasses[sc.GetName()]; found {
-				sc.Annotations[storageutil.IsDefaultStorageClassAnnotation] = "true"
-				_, err := client.StorageV1().StorageClasses().Update(ctx, &sc, metav1.UpdateOptions{})
-				framework.ExpectNoError(err)
-			}
+func setStorageClassDefault(ctx context.Context, client clientset.Interface, scName string, isDefault string) {
+	gomega.Eventually(ctx, func() error {
+		sc, err := client.StorageV1().StorageClasses().Get(ctx, scName, metav1.GetOptions{})
+		if err != nil {
+			return err
 		}
-	}
 
-}
+		if sc.Annotations == nil {
+			sc.Annotations = make(map[string]string)
+		}
+		sc.Annotations[storageutil.IsDefaultStorageClassAnnotation] = isDefault
 
-func waitForPVCStorageClass(ctx context.Context, c clientset.Interface, namespace, pvcName, scName string, timeout time.Duration) (*v1.PersistentVolumeClaim, error) {
-	var watchedPVC *v1.PersistentVolumeClaim
+		_, err = client.StorageV1().StorageClasses().Update(ctx, sc, metav1.UpdateOptions{})
+		return err
+	}).WithPolling(framework.Poll).WithTimeout(framework.ClaimProvisionTimeout).
+		Should(gomega.Succeed(), "Failed to update StorageClass %s", scName)
+}
 
-	err := wait.PollUntilContextTimeout(ctx, 1*time.Second, timeout, false, func(ctx context.Context) (bool, error) {
-		var err error
-		watchedPVC, err = c.CoreV1().PersistentVolumeClaims(namespace).Get(ctx, pvcName, metav1.GetOptions{})
+func ensureNoDefaultStorageClasses(ctx context.Context, client clientset.Interface) {
+	gomega.Eventually(ctx, func() (int, error) {
+		defaultSCs, err := getDefaultStorageClasses(ctx, client)
 		if err != nil {
-			return true, err
+			return 0, err
 		}
+		return len(defaultSCs), nil
+	}).WithPolling(framework.Poll).WithTimeout(2*time.Minute).
+		Should(gomega.BeZero(), "Expected no default StorageClasses")
+}
 
-		if watchedPVC.Spec.StorageClassName == nil {
-			return false, nil // Poll until PVC has correct SC
-		}
-		if watchedPVC.Spec.StorageClassName != nil && *watchedPVC.Spec.StorageClassName != scName {
-			framework.Logf("PersistentVolumeClaim %s has unexpected StorageClass %v, expected StorageClass is: %v", watchedPVC.Name, *watchedPVC.Spec.StorageClassName, scName)
-			return false, nil // Log unexpected SC and continue poll (something might be changing default SC)
-		}
-		return true, nil // Correct SC name found on PVC
-	})
+func createPVC(ctx context.Context, client clientset.Interface, namespace string) *v1.PersistentVolumeClaim {
+	pvcConfig := e2epv.PersistentVolumeClaimConfig{
+		Name: "test-pvc",
+	}
 
-	if err != nil {
-		return watchedPVC, fmt.Errorf("error waiting for claim %s to have StorageClass set to %s: %w", pvcName, scName, err)
+	ginkgo.By("Creating a PVC")
+	pvcObj := e2epv.MakePersistentVolumeClaim(pvcConfig, namespace)
+	var pvc *v1.PersistentVolumeClaim
+	gomega.Eventually(ctx, func() error {
```

Comment:
I'm unsure whether we should use the `Eventually` to wrap `create` or `delete` the PVC object. We often call `e2epv.CreatePVC` and `e2epv.DeletePersistentVolumeClaim` functions immediately in many places. 

/cc @pohly 

Can you help me to review this PR? Thanks a lot!

Reply:
That makes sense - I understood https://github.com/kubernetes/kubernetes/pull/127790#discussion_r1794574457 as `ExpectNoError` to be an anti-pattern, but you are definitely right. I looked at other tests, and the only places were it makes sense to use `Eventually` are for `ensureNoDefaultStorageClasses` and verifying that the PVC is assigned the default SC. Fixed in latest rev.

Reply:
Most tests rely on "simple" API calls to not fail and don't retry. There are some retry mechanisms inside client-go itself which help, so this is typically okay.